### PR TITLE
Hsm2 compatible with hsm3

### PR DESCRIPTION
**Pull request by kelvinator07**

### CODE REVIEW

**Comment by kelvinator07** [comment]
pipeline:run

**Comment by kelvinator07** [comment]
pipeline:run

**Comment by kelvinator07** [comment]
pipeline:run

**Review by jeremy-then — APPROVED** [review]
LGTM.

File: src/test/java/co/rsk/federate/signing/hsm/message/ReleaseCreationInformationGetterTest.java [hunk 1]
```diff
@@ -79,6 +87,17 @@ public void createGetTxInfoToSign_returnOK() throws HSMReleaseCreationInformatio
...
+    public void createGetTxInfoToSign_returnOK_v3(ReleaseCreationInformationGetter information, Keccak256 rskTxHash,
+                                                  BtcTransaction btcTransaction, Block block, TransactionReceipt transactionReceipt) throws HSMReleaseCreationInformationException {
+        ReleaseCreationInformation releaseCreationInformation =
+            information.getTxInfoToSign(3, rskTxHash, btcTransaction);
+
+        Assert.assertEquals(releaseCreationInformation.getBlock(), block);
+        Assert.assertEquals(transactionReceipt, releaseCreationInformation.getTransactionReceipt());
+        Assert.assertEquals(rskTxHash, releaseCreationInformation.getReleaseRskTxHash());
+        Assert.assertEquals(btcTransaction, releaseCreationInformation.getBtcTransaction());
+    }
```
Comment (nathanieliov) [inline src/test/java/co/rsk/federate/signing/hsm/message/ReleaseCreationInformationGetterTest.java:99]: What if you add the version as an argument? That way you only need one method

File: src/test/java/co/rsk/federate/signing/hsm/message/ReleaseCreationInformationGetterTest.java [hunk 2]
```diff
@@ -69,7 +69,15 @@ public void createGetTxInfoToSign_returnOK() throws HSMReleaseCreationInformatio
...
                 blockStore
         );
+        // HSM V2
+        createGetTxInfoToSign_returnOK_v2(information, rskTxHash, btcTransaction, block, transactionReceipt);
 
+        // HSM V3
+        createGetTxInfoToSign_returnOK_v3(information, rskTxHash, btcTransaction, block, transactionReceipt);
+    }
+
+    public void createGetTxInfoToSign_returnOK_v2(ReleaseCreationInformationGetter information, Keccak256 rskTxHash,
```
Comment (nathanieliov) [inline src/test/java/co/rsk/federate/signing/hsm/message/ReleaseCreationInformationGetterTest.java:79]: What if you add the version as an argument? That way you only need one method

**Review by nathanieliov — COMMENTED** [review]
Good job K. Just a minor comment.

File: src/test/java/co/rsk/federate/signing/hsm/message/ReleaseCreationInformationGetterTest.java (hunk 2 above)
Comment (kelvinator07) [reply src/test/java/co/rsk/federate/signing/hsm/message/ReleaseCreationInformationGetterTest.java]: Thanks Nath, Resolved

File: src/test/java/co/rsk/federate/signing/hsm/client/HSMClientVersion2BTCTest.java [hunk 3]
```diff
@@ -445,6 +437,34 @@ public void updateAncestorBlock_ok() throws HSMClientException, JsonRpcException
...
+
+        BlockHeader blockHeader = mock(BlockHeader.class);
+        when(blockHeader.getEncoded(true, false)).thenReturn(new byte[]{});
+
+        client.updateAncestorBlock(new UpdateAncestorBlockMessage(Arrays.asList(blockHeader, blockHeader, blockHeader)));
+        ArgumentCaptor<JsonNode> captor = ArgumentCaptor.forClass(JsonNode.class);
+        verify(jsonRpcClient, times(3)).send(captor.capture());
+        List<JsonNode> capturedArguments = captor.getAllValues();
+        Assert.assertEquals("blockchainState", capturedArguments.get(0).get("command").asText());
+        Assert.assertEquals("updateAncestorBlock", capturedArguments.get(1).get("command").asText());
```
Comment (marcos-iov) [inline src/test/java/co/rsk/federate/signing/hsm/client/HSMClientVersion2BTCTest.java:462]: Why is updateAncestorBlock called twice?

Comment (kelvinator07) [reply src/test/java/co/rsk/federate/signing/hsm/client/HSMClientVersion2BTCTest.java:462]: The BlockHeaders Was Broken Into 2 Chunks.. Hence, Calling The updateAncestorBlock Twice

File: src/test/java/co/rsk/federate/signing/hsm/client/HSMClientProviderTest.java [hunk 4]
```diff
@@ -71,6 +71,17 @@ public void getClientV2MST() throws Exception {
         Assert.assertTrue(client instanceof HSMClientVersion2RskMst);
     }
 
+    @Test
+    public void getClientV3() throws Exception {
+        HSMClientProtocol protocol = mock(HSMClientProtocol.class);
+        when(protocol.getVersion()).thenReturn(3);
+
+        HSMClientProvider clientProvider = new HSMClientProvider(protocol, "BTC");
```
[the remark this answers is not in the document]
Comment (kelvinator07) [reply src/test/java/co/rsk/federate/signing/hsm/client/HSMClientProviderTest.java:79]: Resolved

File: src/test/java/co/rsk/federate/signing/hsm/message/AdvanceBlockchainMessageTest.java [hunk 5]
```diff
@@ -5,6 +5,7 @@
 import org.junit.Test;
 import org.spongycastle.util.encoders.Hex;
 
+import java.util.ArrayList;
```
[the remark this answers is not in the document]
Comment (kelvinator07) [reply src/test/java/co/rsk/federate/signing/hsm/message/AdvanceBlockchainMessageTest.java]: No, Removed

File: src/test/java/co/rsk/federate/signing/hsm/message/ReleaseCreationInformationGetterTest.java [hunk 6]
```diff
@@ -69,9 +66,21 @@ public void createGetTxInfoToSign_returnOK() throws HSMReleaseCreationInformatio
...
                 blockStore
         );
+        // HSM V2
+        createGetTxInfoToSign_returnOK(information, rskTxHash, btcTransaction, block, transactionReceipt, 2);
 
+        // HSM V3
+        createGetTxInfoToSign_returnOK(information, rskTxHash, btcTransaction, block, transactionReceipt, 3);
+    }
+
+    public void createGetTxInfoToSign_returnOK(ReleaseCreationInformationGetter information,
```
[the remark this answers is not in the document]
Comment (kelvinator07) [reply src/test/java/co/rsk/federate/signing/hsm/message/ReleaseCreationInformationGetterTest.java]: Resolved

File: src/test/java/co/rsk/federate/signing/hsm/client/HSMClientVersion2BTCTest.java (hunk 3 above)
Comment (marcos-iov) [reply src/test/java/co/rsk/federate/signing/hsm/client/HSMClientVersion2BTCTest.java:462]: I see, thanks. Might be worth adding a comment then. Or changing the test name

**Review by nathanieliov — APPROVED** [review]

File: src/test/java/co/rsk/federate/signing/hsm/client/HSMClientVersion2BTCTest.java (hunk 3 above)
Comment (kelvinator07) [reply src/test/java/co/rsk/federate/signing/hsm/client/HSMClientVersion2BTCTest.java:463]: Okay, Done.

**Review by marcos-iov — APPROVED** [review]

**Pull request merged**